redis-benchmark: add tests, --version, a minor bug fixes

filipecosta90 opened the pull request:
## added tests

This PR is a maintenance one for redis-benchmark, adding the following tests for standalone redis:
- full test suite
- multi-thread full test suite
- pipelined commands full test suite
- arbitrary command

## added --version
Apart from it, and to make it easier to reference redis-benchmark version it applies the same version format as seen on redis-cli. 
```
$ redis-benchmark --version
redis-benchmark 255.255.255 (git:7223d1d0)
```

oranagra commented:
@filipecosta90 thank you for this.
I see the additional test takes 45 seconds, since we're not really looking doing benchmark (don't care about the performance results), i would like to find a way to trim it down to just a few seconds.

sine the test names on success / error are printed without context, i think it would be better to add a `benchmark:` prefix to the individual tests.

it's nice that we now have coverage to make sure redis benchmark doesn't crash, but i think we may want to do a bit more to validate that the output on success seems correct.

filipecosta90 commented:
Tks for the initial review Oran. Will proceed with the changes and report back :+1:

filipecosta90 commented:
@oranagra I've changed the naming and added further clauses as per the above comment. 
Apart from it while adding more tests and proper checks on each test I've noticed that redis-benchmark with pipeline is not enforcing upper bound to number of requests. I've opened an issue describing the problem https://github.com/redis/redis/issues/7960 and added the the fix and more tests to this PR.

oranagra commented on src/redis-benchmark.c:
```diff
@@ -577,7 +594,7 @@ static void writeHandler(aeEventLoop *el, int fd, void *privdata, int mask) {
         /* Enforce upper bound to number of requests. */
         int requests_issued = 0;
-        atomicGetIncr(config.requests_issued, requests_issued, 1);
+        atomicGetIncr(config.requests_issued, requests_issued, config.pipeline);
```
@filipecosta90 are you sure that's right? every write carries this count of commands?
i'm not familiar with the redis_benchmark code too good. doesn't it use a streaming pipeline? i.e. one that initially sends N commands, and then as soon as getting the response for one, it issues another one, so that at any given moment there are N commands in flight?

even f not, i see that we have a `c->pending` variable, maybe that's what we should be using? (rather than assuming each write has N commands)

please look into it and tell me what you found.

filipecosta90 replied:
> even f not, i see that we have a `c->pending` variable, maybe that's what we should be using? (rather than assuming each write has N commands)

@oranagra looking at [src/redis-benchmark.c#L716](https://github.com/redis/redis/blob/unstable/src/redis-benchmark.c#L716) we can check that c->obuf has always N pipeline commands. 

Regarding using c->pending istead of config.pipeline I believe we might get into an issue due to the initial setup commands `c->pending = config.pipeline+c->prefix_pending;` . Apart from it, and given that we only enter this condition when `c->written == 0` the value of c->pending inside that condition will always be `c->pending = config.pipeline` due to https://github.com/redis/redis/blob/unstable/src/redis-benchmark.c#L381;

oranagra replied:
ok...
so that that i'm sure we're on the same line.
it creates a client and an obuf with a single set of pipeline commands.
then writes to the socket, disables the write handler, enables the read handler.
when it's done reading responses it "re-sets" the client, enables the write handler and run another iteration.
so working in bursts rather than maintaining a constant in-flight pipeline.

so now i see that `clientDone` (which is called by the read handler) does this logic too:
```
    if (requests_finished >= config.requests) {
        freeClient(c);
```
so i wonder why it is needed in the write handler.
in theory it's wrong for the write handler to free the client (if it writes anything we wanna wait for the responses), but i suppose that's just dead code (the read handler will free the client rather than re-set it).

can you please look into that?

filipecosta90 replied:
> so now i see that clientDone (which is called by the read handler) does this logic too:

>    if (requests_finished >= config.requests) {
        freeClient(c); 

> so i wonder why it is needed in the write handler.
in theory it's wrong for the write handler to free the client (if it writes anything we wanna wait for the responses), but i suppose that's just dead code (the read handler will free the client rather than re-set it).

> can you please look into that?

@oranagra agree. It's wrong to free the client in the write handler given the chance of the replies still to be processed not being processed at all. Confirmed that [removing the freeClient within the writeHandler](https://github.com/redis/redis/pull/7947/files#diff-f37c3eacb5f6ac27e8fa820f102e8b57f0f33c7b46e5ef2ba46b8ad8a56c86a6L582) does not affect the tests and redis-benchmark still works as expected.

oranagra reviewed (commented):
@filipecosta90 i see the test still takes some 25 seconds.
please try to shave it down further.

i see most tests only run some 1000 commands, so maybe it's the `keyspace length` test that's taking most time.

filipecosta90 commented:
> @filipecosta90 i see the test still takes some 25 seconds.
> please try to shave it down further.
> 
> i see most tests only run some 1000 commands, so maybe it's the `keyspace length` test that's taking most time.

Thank you for the quick review @oranagra. I've tried to nail down further the tests duration and now running solely the redis-benchmark locally it's showming 1 sec duration:
```
$ tclsh tests/test_helper.tcl --verbose --single integration/redis-benchmark 
Cleanup: may take some time... OK
Starting test server at port 11111
[ready]: 657206
Testing integration/redis-benchmark
[ready]: 657205
[ready]: 657207
[ready]: 657208
[ready]: 657209
[ready]: 657210
[ready]: 657211
[ready]: 657212
[ready]: 657213
[ready]: 657214
[ready]: 657215
[ready]: 657216
[ready]: 657217
[ready]: 657218
[ready]: 657219
[ready]: 657220
=== (benchmark) Starting server 127.0.0.1:21611 ok
=== () Starting server 127.0.0.1:21612 ok
[ok]: benchmark: set,get
[ok]: benchmark: full test suite
[ok]: benchmark: multi-thread set,get
[ok]: benchmark: pipelined full set,get
[ok]: benchmark: arbitrary command
[ok]: benchmark: keyspace length
[1/1 done]: integration/redis-benchmark (1 seconds)

                   The End

Execution time of different units:
  1 seconds - integration/redis-benchmark

\o/ All tests passed without errors!

Cleanup: may take some time... OK

``` 
Can you check on your side if it also takes only that?

oranagra commented on tests/integration/redis-benchmark.tcl:
```diff
@@ -93,13 +93,13 @@ start_server {tags {"benchmark"}} {
             r config resetstat
             r flushall
-            set cmd [redisbenchmark $master_host $master_port "-c 5 -n 1000 -e INCRBYFLOAT mykey 10.0"]
+            set cmd [redisbenchmark $master_host $master_port "-c 5 -n 150s -e INCRBYFLOAT mykey 10.0"]
```
is the `s` in `150s` a typo?

filipecosta90 replied:
yes. fixing it

oranagra reviewed (commented):
ohh, so it's probably just the slowness of the github actions.
up to 5 or 10 seconds on your local computers seems fine (if you feel you rather roll some of these changes back that's ok with me.)

oranagra commented:
@filipecosta90 looks like this test it failing when the tests are executed with `--tls`
https://github.com/redis/redis/pull/7965/checks?check_run_id=1309191804
i guess your follow up PR solves it?

filipecosta90 commented:
> @filipecosta90 looks like this test it failing when the tests are executed with `--tls`
> https://github.com/redis/redis/pull/7965/checks?check_run_id=1309191804
> i guess your follow up PR solves it?

@oranagra yes [here](https://github.com/redis/redis/pull/7959/files#diff-2f3e48cc18abe0ce258bebcab54936f297b23099f200785d3a71026f944f06a7R1).
 Do you want to separate this fix from that PR? or no need?

oranagra commented:
@filipecosta90 how can that fix the failing tests? (without adding TLS to redis-benchmark, or skipping the tests in that mode).
